pkg/package_archive/lib/package_archive.dart

```diff
@@ -0,0 +1,229 @@
+    errors.add(ArchiveError('pubspec.yaml is missing.'));
+    return PackageArchive(errors: errors);
+  }
+
+  final pubspecContent = await readTarballFile(archivePath, 'pubspec.yaml');
+  // Large pubspec content should be rejected, as either a storage limit will be
+  // limiting it, or it will slow down queries and processing for very little
+  // reason.
+  if (pubspecContent.length > 128 * 1024) {
+    errors.add(ArchiveError('pubspec.yaml is too large.'));
+  }
+  Pubspec pubspec;
+  try {
+    pubspec = Pubspec.parse(pubspecContent);
+  } catch (e) {
+    pubspec = Pubspec.parse(pubspecContent, lenient: true);
+    errors.add(ArchiveError('Error parsing pubspec.yaml: $e'));
+  }
+
+  // Check whether the files can be extracted on case-preserving file systems
+  // (e.g. on Windows). We can't allow two files with the same case-insensitive
+  // name.
+  final lowerCaseFiles = <String, List<String>>{};
+  for (String file in files) {
+    final lower = file.toLowerCase();
+    lowerCaseFiles.putIfAbsent(lower, () => <String>[]).add(file);
+  }
+  final fileNameCollisions =
+      lowerCaseFiles.values.firstWhere((l) => l.length > 1, orElse: () => null);
+  if (fileNameCollisions != null) {
+    errors.add(ArchiveError(
+        'Filename collision on case-preserving file systems: ${fileNameCollisions.join(' vs. ')}.'));
+  }
+
+  if (pubspec.name == null || pubspec.name.trim().isEmpty) {
+    errors.add(ArchiveError('pubspec.yaml is missing `name`.'));
+    return PackageArchive(errors: errors);
+  }
+  if (pubspec.version == null) {
+    errors.add(ArchiveError('pubspec.yaml is missing `version`.'));
+  }
+
+  final package = pubspec.name;
+
+  Future<String> extractContent(String contentPath) async {
+    if (contentPath == null) return null;
+    final content = await readTarballFile(archivePath, contentPath,
+        maxLength: _maxStoredLength);
+    if (content != null && content.trim().isEmpty) {
+      return null;
+    }
+    return content;
+  }
+
+  String readmePath = searchForFile(readmeFileNames);
+  final readmeContent = await extractContent(readmePath);
+  if (readmeContent == null) {
+    readmePath = null;
+  }
+
+  String changelogPath = searchForFile(changelogFileNames);
+  final changelogContent = await extractContent(changelogPath);
+  if (changelogContent == null) {
+    changelogPath = null;
+  }
+
+  String examplePath = searchForFile(exampleFileCandidates(package));
+  final exampleContent = await extractContent(examplePath);
+  if (exampleContent == null) {
+    examplePath = null;
+  }
+
+  final libraries = files
+      .where((file) => file.startsWith('lib/'))
+      .where((file) => !file.startsWith('lib/src'))
+      .where((file) => file.endsWith('.dart'))
+      .map((file) => file.substring('lib/'.length))
+      .toList();
+
+  errors.addAll(validatePackageName(pubspec.name));
+  errors.addAll(syntaxCheckHomepageUrl(
+      pubspec.homepage ?? pubspec.repository?.toString()));
```

Comment:
By increasing the level of abstraction I mean... we probably need something like:
https://github.com/dart-lang/pub/tree/master/lib/src/validator

So, maybe we have a `Package` class which has a function to list and read files... and long with parsed pubspec, and README, etc...

We would then have a long list of `Validators` that are passed a `Package` instance, and then return a list of `Issues`.

This way, each `Validator` is probably just a function... we can have them in separate files, and we can easily switch on/off which ones are active by adding removing the functions to/from a list of functions.

Just as importantly, these validators can be tested independently, and each one of them would deal with a separate and independent concern. One for `syntaxCheckHomepageUrl` another for `validatePackageName`.... maybe a different one to `pingHomepageUrl`, etc...

Reply:
Agreed, but let's do it in a separate PR.